doc/developer/design/20230227_external_introspection.md

```diff
@@ -0,0 +1,175 @@
+shown below.
+
+![diagram of proposed feature](fg.png)
+
+## Per-query metadata
+
+We should collected the following metadata for each query:
+
+* The timestamp at which the query executed
+* The frontiers of all dependencies
+* The optimized and physical query plans
+* The ID of the dataflow (if it is not a simple peek) used to service
+  the query
+* The SQL text of the query
+
+We can then save all this information in a table with non-trivial
+retention, and surface it from the web UI.
+
+## User-friendly rendering of query plans
+
+Currently we have text-only `EXPLAIN PLAN` output. We should render
+this data in visual form, as a graph. We should also (whenever
+possible) flow column names from the source relations through the
+nodes, so that we can show something more useful than `#0`, etc.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+## Data sources
+All features will be implemented in the web UI using React, querying
+the `mz_internal` relations for the necessary data. Relations that
+will be used include:
+
+* For the hierarchical dataflow visualizer:
+    * `mz_dataflow_operators`
+    * `mz_dataflow_channels`
+    * `mz_arrangement_sizes`
+    * `mz_dataflow_addresses`
+* For the global frontier lag visualizer:
+    * `mz_object_dependencies`
+    * `mz_cluster_replica_frontiers`
+* For the flamegraph views: same as the hierarchical dataflow
+  visualizer
+* For the per-query metadata: New table to be created
+  (`mz_query_metadata`).
+* For the user-friendly plan rendering: parsed `EXPLAIN PLAN` output
+  (possibly from `mz_query_metadata`, or entered manually by the user).
+
+## Layout and rendering
+
+We will use the [d3-graphviz](https://github.com/magjac/d3-graphviz)
+library for layout and rendering of the hierarchical dataflow
+visualizer, global frontier lag visualizer, and user-friendly plan
+rendering. We will use
+[d3-flame-graph](https://github.com/spiermar/d3-flame-graph) -- which
+we are already using on the internal side -- for the flamegraph
+visualizer.
+
+# Rollout and Lifecycle
+
+We will be considering this an experimental feature until such time as
+we can involve professional product designers and front-end
+engineers. Until that time, the level of polish of the UX may not
+reach the same standards as the rest of the site, so we will require
+users to click a link with a label like "Advanced Features" or similar
+in order to access the tools.
+
+We will also use LaunchDarkly to gate access to the feature, and not
+launch it at all to the public until we have gotten some internal
+feedback from support and DevEx that it is useful.
+
+We will have a separate LaunchDarkly flag just for the
+`mz_query_metadata` table, since this will have potentially large cost
+in high-QPS scenarios.
+
+## Testing and observability
+
+(TBD -- will cover this in our meeting with Robin tomorrow and get an
+overview from him of the testing strategy for this kind of feature)
```

Comment:
Yes, I agree. I haven't filled out this section yet because I'm not sure what frameworks/tools are used for UI testing now, but we definitely want good end-to-end test coverage.